crates/snforge-scarb-plugin/src/config_statement.rs

```diff
@@ -0,0 +1,162 @@
+use crate::{
+    args::Arguments,
+    asserts::assert_is_used_once,
+    attributes::AttributeCollector,
+    parse::{parse, parse_args},
+};
+use cairo_lang_macro::{Diagnostic, Diagnostics, ProcMacroResult, TokenStream};
+use cairo_lang_syntax::node::{
+    ast::{Condition, Expr, FunctionWithBody, GenericArg, GenericArgValue, Statement},
+    db::SyntaxGroup,
+    helpers::{GetIdentifier, PathSegmentEx},
+    TypedSyntaxNode,
+};
+use cairo_lang_utils::Upcast;
+use indoc::formatdoc;
+
+#[allow(clippy::needless_pass_by_value)]
+pub fn extend_with_config_cheatcodes<Collector>(
+    args: TokenStream,
+    item: TokenStream,
+) -> ProcMacroResult
+where
+    Collector: AttributeCollector,
+{
+    match extend_with_config_cheatcodes_internal::<Collector>(&args, &item) {
+        Ok((item, warn)) => {
+            let result = ProcMacroResult::new(TokenStream::new(item));
+
+            if let Some(warn) = warn {
+                result.with_diagnostics(warn.into())
+            } else {
+                result
+            }
+        }
+        Err(diagnostics) => ProcMacroResult::new(item).with_diagnostics(diagnostics),
+    }
+}
+
+fn extend_with_config_cheatcodes_internal<Collector>(
+    args: &TokenStream,
+    item: &TokenStream,
+) -> Result<(String, Option<Diagnostic>), Diagnostics>
+where
+    Collector: AttributeCollector,
+{
+    let item = item.to_string();
+    let (db, func) = parse::<Collector>(&item)?;
+
+    let db = db.upcast();
+
+    assert_is_used_once::<Collector>(db, &func)?;
+
+    let (args_db, args) = parse_args::<Collector>(&args.to_string())?;
+
+    let (args, empty_args_list_warn) = Arguments::new::<Collector>(args_db.upcast(), args);
+
+    let value = Collector::args_into_body(args_db.upcast(), args).map_err(|err| {
+        if let Some(empty_args_list_warn) = &empty_args_list_warn {
+            err.warn(&empty_args_list_warn.message)
+        } else {
+            err
+        }
+    })?;
+
+    let cheatcode_name = Collector::CHEATCODE_NAME;
+
+    let config_cheatcode = formatdoc!(
+        r#"
+            let mut data = array![];
+
+            {value}
+            .serialize(ref data);
+
+            cheatcode::<'{cheatcode_name}'>(data);
+        "#
+    );
+
+    Ok((
+        append_config_statements(db, &func, &config_cheatcode),
+        empty_args_list_warn,
+    ))
+}
+
+const CONFIG_CHEATCODE: &str = "was_configuration_set";
+
+macro_rules! propagate {
+    ($pattern:pat = $expression:expr) => {
+        let $pattern = $expression else { return None };
+    };
+}
```

Comment:
Nitpick: from an external code reader POV, this macro decreases readability. I had to check its source to learn what it does, while the original code is quite similar in characters size.

Reply:
Good point.